feat: add timelocks

sajanrajdev opened the pull request:
The following changes were made:
- A wrapper to the OZ TimelockController was implemented using the AccessControlEnumberable instead of the regular AccessControl library to enable better roles visibility and management
- A deployment and basic functionality test suite was introduced for the timelocks
- The timelocks deployment, their configuration and the system's governance configuration were added to the deployment script  according to the governance [specifications](https://www.google.com/url?sa=t&rct=j&esrc=s&source=appssearch&uact=8&cd=0&cad=rja&q&sig2=futBE7gpvOt_lnF2B-WOhw&ved=0ahUKEwjc3riIsa-CAxUonAoJHcCiAz04ABABKAAwAA&url=https://drive.google.com/a/badger.com/open?id%3D1ZMo857wHmx7KMudCooIW5cvPf9nZr9Dz7XSSXT4066s&usg=AOvVaw2yFi3I6UttlYB5mcRVkacu).
- The fee recipient was set to the `feeRecipientOwner` (multisig) as the default in the deployment script. We will explore switching into a more elaborate soution in the future. For the time being, the fee recipient contract is still being deployed.

Note for reviewers: Please confirm that no governable function signature was left out.

rayeaster commented on packages/contracts/mainnetDeployment/governanceSignatures.js:
```diff
@@ -0,0 +1,50 @@
+// Define the permissioned function signatures
+const SET_STAKING_REWARD_SPLIT_SIG = (ethers.utils.id('setStakingRewardSplit(uint256)')).substring(0, 10);
```
it might help to document the contract for each function and `setMaxFeeBps()` seems already removed

```
// CdpManager
const SET_STAKING_REWARD_SPLIT_SIG = ...
...
// CdpManagerStorage
const SET_GRACE_PERIOD_SIG = ...
...
// EBTCToken
const MINT_SIG = ...
...
// PriceFeed
const SET_FALLBACK_CALLER_SIG = ...
...
// BorrowerOperations & ActivePool
const SET_FEE_BPS_SIG = ...
...
// ActivePool & CollSurplusPool
const SWEEP_TOKEN_SIG = ...
...
// ActivePool
const CLAIM_FEE_RECIPIENT_COLL_SIG = ...
...
// BorrowerOperations & ActivePool
const SET_FEE_RECIPIENT_ADDRESS_SIG = ...
...
// Governor
const SET_ROLE_NAME_SIG = ...
...
// RolesAuthority
const SET_USER_ROLE_SIG = ...
...
// Ownable 
const TRANSFER_OWNERSHIP_SIG = ...
...
// AuthNoOwner & Auth
const SET_AUTHORITY_SIG = ...
...
```

sajanrajdev replied:
Good call, addressed.

rayeaster commented on packages/contracts/foundry_test/TimelockOperations.t.sol:
```diff
@@ -0,0 +1,312 @@
+    /// @dev Assumes default governance still has ownerships and is pranked
+    function _grantAllGovernorCapabilitiesToRole(uint8 role) internal {
+        // List of all setter function signatures in Governor contract
+        bytes4[] memory funcSigs = new bytes4[](9);
```
the `funcSigs` has a length of `9` but following assignment seems miss index `3` and `5` ?

sajanrajdev replied:
Good catch

rayeaster commented on packages/contracts/mainnetDeployment/eBTCDeployScript.js:
```diff
@@ -266,6 +300,184 @@ class EBTCDeployerScript {
+        await authority.setRoleCapability(5, coreContracts.activePool.address, govSig.SET_FEE_BPS_SIG, true);
+        await authority.setRoleCapability(5, coreContracts.activePool.address, govSig.SET_FLASH_LOANS_PAUSED_SIG, true);
+        await authority.setRoleCapability(5, coreContracts.activePool.address, govSig.SET_FEE_RECIPIENT_ADDRESS_SIG, true);
+        await authority.setRoleCapability(6, coreContracts.activePool.address, govSig.SWEEP_TOKEN_SIG, true);
```
`CollSurplusPool` also got a `sweepToken` function

sajanrajdev replied:
Yea, good call. Will add.

rayeaster commented on packages/contracts/mainnetDeployment/eBTCDeployScript.js:
```diff
@@ -185,6 +209,16 @@ class EBTCDeployerScript {
+        let _constructorArgs = [this.highSecDelay, [this.ecosystemMultisig], [this.ecosystemMultisig], this.highSecAdmin]
+        this.highSecTimelock = await this.deployOrLoadState(HIGHSEC_TIMELOCK_STATE_NAME, [], _constructorArgs)
+
+        console.log(chalk.cyan("[HighSecTimelock]"))
```
should be `"[LowSecTimelock]"`?

sajanrajdev replied:
Good catch, addressed.

rayeaster reviewed (approved):
LGTM to add an OZ timelock-based governance architecture

sajanrajdev commented:
There were a few issues that showed up when running against Goerli:
- Deployment and verification of the Timelocks wasn't working properly. This was due to the the way the OZ libs were being imported. Switched to the conventional HH method and modified the Foundry remappings to use this format. - Fixed
- Sequential transactions were not waiting for the previous to complete before submitting the next one. Added tx.wait() - Fixed
- Added some mappings for the roles, roleNames, targets and users to make the governance wiring cleaner and programmatic. In the same way, this allows better logging and deployment feedback. - Fixed.

sajanrajdev commented:
Small changes introduced:
- HighSec timelock won't have ownership, instead it is assigned all roles except for mint/burning - Nobody should have mint/burn access unless explicitly assigned to them
- LowSec gets sweeping and coll claiming for the fee recipient (as per design and in case it is needed - unlikely)
- Authority's Owernship transfer is commented out. The deployer will burn ownership after deployment and once everything is trippled checked.

GalloDaSballo commented:
> * LowSec gets sweeping and coll claiming for the fee recipient (as per design and in case it is needed - unlikely)

Is this documented in a doc + checklist?

Is the deployment sequence tested / testable?

sajanrajdev commented:
> > * LowSec gets sweeping and coll claiming for the fee recipient (as per design and in case it is needed - unlikely)
> 
> Is this documented in a doc + checklist?
Here is the doc for reference: https://docs.google.com/document/d/1ZMo857wHmx7KMudCooIW5cvPf9nZr9Dz7XSSXT4066s/edit
What do you mean by checklist?

> Is the deployment sequence tested / testable?
Yes, I tested it in Goerli already and it worked with no issues. It can be tested against local or any testnet

dapp-whisperer commented on yarn.lock:
```diff
@@ -1212,12 +1212,7 @@
     find-up "^4.1.0"
     fs-extra "^8.1.0"
 
-"@openzeppelin/contracts@^3.3.0":
```
@GalloDaSballo wanted to call out this dependency change

GalloDaSballo replied:
These have known vulnerabilities:
https://security.snyk.io/vuln?search=OpenZeppelin%2Fcontracts

GalloDaSballo replied:
We should use: https://www.npmjs.com/package/@openzeppelin/contracts/v/4.9.5

GalloDaSballo replied:
Well known, bug fixed, still supported

dapp-whisperer replied:
thanks, fixed version @ 4.9.5

dapp-whisperer commented:
we will need to add primary oracle governance to HIGHSEC, but need to merge into that branch first.
